WatchApp Extension/Controllers/ActionHUDController.swift

```diff
@@ -78,31 +79,77 @@ final class ActionHUDController: HUDInterfaceController {
         }
     }
 
+    private func updateForPreMeal(enabled: Bool) {
+        if enabled {
+            preMealButtonGroup.state = .on
+        } else {
+            preMealButtonGroup.turnOff()
+        }
+    }
+
     private func updateForOverrideContext(_ context: TemporaryScheduleOverride.Context?) {
         switch context {
         case nil:
-            preMealButtonGroup.turnOff()
             overrideButtonGroup.turnOff()
-        case .preMeal?:
-            preMealButtonGroup.state = .on
-            overrideButtonGroup.turnOff()
-        case .legacyWorkout?, .preset?, .custom?:
+        case .preset?, .custom?:
+            overrideButtonGroup.state = .on
+        case .legacyWorkout?:
             preMealButtonGroup.turnOff()
             overrideButtonGroup.state = .on
+        case .preMeal?:
+            assertionFailure()
         }
     }
 
     // MARK: - Menu Items
 
+    private var pendingMessageResponses = 0
+
     @IBAction func togglePreMealMode() {
-        let override: TemporaryScheduleOverride?
-        if preMealButtonGroup.state == .on {
-            override = nil
+        let isPreMealEnabled = preMealButtonGroup.state == .off
+        updateForPreMeal(enabled: isPreMealEnabled)
+        pendingMessageResponses += 1
+
+        var settings = loopManager.settings
+        if isPreMealEnabled {
+            settings.enablePreMealOverride(for: .hours(1))
         } else {
-            override = loopManager.settings.preMealOverride(for: .hours(1))
+            settings.clearOverride(matching: .preMeal)
         }
 
-        sendOverride(override)
+        let userInfo = LoopSettingsUserInfo(settings: settings)
+        do {
+            try WCSession.default.sendSettingsUpdateMessage(userInfo, completionHandler: { (result) in
+                DispatchQueue.main.async {
+                    self.pendingMessageResponses -= 1
+
+                    switch result {
+                    case .success(let context):
+                        if self.pendingMessageResponses == 0 {
+                            self.loopManager.settings.preMealOverride = settings.preMealOverride
+                        }
+
+                        ExtensionDelegate.shared().loopManager.updateContext(context)
+                    case .failure(let error):
+                        if self.pendingMessageResponses == 0 {
+                            ExtensionDelegate.shared().present(error)
+                            self.updateForPreMeal(enabled: isPreMealEnabled)
+                        }
+                    }
+                }
+            })
+        } catch {
+            pendingMessageResponses -= 1
+            if pendingMessageResponses == 0 {
+                updateForPreMeal(enabled: isPreMealEnabled)
+            }
+            presentAlert(
```

Comment:
Also, will this catch always occur on `DispatchQueue.main`? If not, might need to enforce that or otherwise could get into a threading issue with `pendingMessageResponses`.

Reply:
> Also, will this catch always occur on `DispatchQueue.main`? If not, might need to enforce that or otherwise could get into a threading issue with `pendingMessageResponses`.

Yes, `catch` clauses always happen on the same queue as their corresponding `do`; this function is called on the main queue in response to a button tap.